### DISCUSSION TITLE
Docker login

### DISCUSSION THREAD

**Discussion by jokay** [renovatebot/renovate#10951] [Request Help] [answered]
As follow-up of the issue [#30](https://gitlab.com/renovate-bot/renovate-runner/-/issues/30#note_634496341).

Is there a way to have the renovate-runner doing a `docker login` prior searching for updated images?

This is the relevant part in `.gitlab-ci.yml` we use:

```yml
renovate:
  stage: renovate
  image: renovate/renovate:slim
  variables:
    RENOVATE_BASE_DIR: ...
    RENOVATE_ENDPOINT: ...
    RENOVATE_PLATFORM: ...
    RENOVATE_GIT_AUTHOR: ...
    RENOVATE_REPOSITORY_CACHE: ...
    RENOVATE_REQUIRE_CONFIG: ...
    RENOVATE_ONBOARDING: ...
    RENOVATE_AUTODISCOVER: ...
    RENOVATE_TOKEN: ...
    GITHUB_COM_TOKEN: ...
  script:
    - renovate ${RENOVATE_EXTRA_FLAGS}
```

As mentionned already tried (without success):

```yml
  script:
    - docker login -u {username} -p {token}
    - renovate ${RENOVATE_EXTRA_FLAGS}
    - docker logout
```

**Comment by jokay** [comment, marked as answer]
Env params `DOCKER_USERNAME` and `DOCKER_PASSWORD` may be used to achieve this.

Source: [config-help/#164](https://github.com/renovatebot/config-help/issues/164)

**Reply by viceice** [reply]
See https://docs.renovatebot.com/private-modules/#hostrules-configuration-using-environment-variables

It's documented.

**Reply by ChadAragorn** [reply]
That link doesn't work @viceice. I found how to use the DOCKER_AUTH_CONFIG env variable you can set in the gitlab ci/cd variable section to authenticate. I had to edit the renovate.json and add the following:

`  "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "docker.io",
      "token": "process.env.DOCKER_AUTH_CONFIG"
    }
  ],`
  
  I don't know that it was the correct way to do this, but it did work.

**Reply by viceice** [reply]
Your config looks wrong `process.env.xxx` only works in `config.js` and only without quotes.



New link, page was moved https://docs.renovatebot.com/getting-started/private-packages/#hostrules-configuration-using-environment-variables

**Reply by ChadAragorn** [reply]
@viceice Maybe it is only supposed to work that way but I can confirm that it works as I said. I placed it inside the renovate.json which is in the base of the project I would like renovated. I tried it first without the quotes, and ended up having this in the debug logs:

> DEBUG: Found renovate.json config file (repository=it/simple-react-fullstack)
 INFO: Repository has invalid config (repository=it/simple-react-fullstack)
       "error": {
         "validationSource": "renovate.json",
         "validationError": "Invalid JSON (parsing failed)",
         "validationMessage": "Syntax error near en\": proce",
         "message": "config-validation",
         "stack": "Error: config-validation\n    at checkForRepoConfigError (/usr/src/app/node_modules/renovate/lib/workers/repository/init/merge.ts:155:17)\n    at mergeRenovateConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/init/merge.ts:176:3)\n    at getRepoConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/init/config.ts:13:12)\n    at initRepo (/usr/src/app/node_modules/renovate/lib/workers/repository/init/index.ts:31:12)\n    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:44:14)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:122:7)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"
       }
       
I wasn't sure exactly what the problem was but I knew it was the line so I added the quotes and then I was able to authenticate to dockerhub just fine using the DOCKER_AUTH_CONFIG variable. The pipeline completed, it spun up a dashboard of stuff to fix for my other project, and it started feature branch pipelines for the changes.

Which project would I put the config.js in, the renovate-runner or the project I want renovated? FYI I am using a renovate-runner on my own self hosted gitlab server.

**Reply by viceice** [reply]
Environment variables are not supported in repo `renovate.json`, so it simply using `process.env.DOCKER_AUTH_CONFIG` as your token.

So i assume that token isn't used and renovate makes anonymous requests. Otherwise is should fail with auth issues.

The `config.js` needs to be placed in your runner repo or passed as file env `RENOVATE_CONFIG_FILE` there.

**Comment by viceice** [comment]
You have two things to do:

1. manual docker login from renovate slim container to allow renovate pulling sidecar container images with credentials
2. add a host rule for docker hub with credentials, so renovate docker datasource can use the credentials too.

**Reply by malmor** [reply]
Are there any additional settings that need to be present in order for this to work?
We are getting some "Too many requests" errors from hub.docker.com even though we are following those two steps.
I looked around at https://docs.renovatebot.com/docker/#registry-authentication but didn't spot any settings that we are missing.

Our setup:

**.gitlab-ci.yml**
```
renovate:
  stage: renovate
  image: renovate/renovate:26.9.0-slim@sha256:10625c808ffc055a37d80c12eb03a3590d3895873847af0d79d6f261bc2f1480
  tags:
    - docker-executor
  services:
    - name: docker:20.10.8-dind@sha256:c0aac52a0f704627c01dfd5177f3d0e5d9bf6e64f2b8a2c35b5b5087ebe94cfc
      command: ["--registry-mirror", "https://registrymirror.company.local"] 
  before_script:
    # get private registry credentials
    - export DOCKERHUB_USERNAME="$(vault kv get ...)"
    - export DOCKERHUB_PASSWORD="$(vault kv get ...)"
    # get public registry credentials
    - export DOCKER_PUBLIC_REGISTRY_USERNAME=$(vault kv get ...)
    - export DOCKER_PUBLIC_REGISTRY_PASSWORD=$(vault kv get ...)
    # configure renovate settings
    - export REDIS_URL=$(vault kv get ...)
    - export RENOVATE_TOKEN=$(vault kv get ...)
    - export GITHUB_COM_TOKEN=$(vault kv get ...)
    # docker login (pull rate limit)
    - echo "${DOCKER_PUBLIC_REGISTRY_PASSWORD}" | docker login --username "${DOCKER_PUBLIC_REGISTRY_USERNAME}" --password-stdin
  script:
    - renovate $RENOVATE_EXTRA_FLAGS
```

**runner config.json**
```
module.exports = {
  hostRules: [
    {
      hostType: "docker",
      matchHost: "https://index.docker.io/",
      username: process.env.DOCKER_PUBLIC_REGISTRY_USERNAME,
      password: process.env.DOCKER_PUBLIC_REGISTRY_PASSWORD,
      timeout: 30000
    },
    {
      hostType: "docker",
      matchHost: "https://registry.company.local",
      username: process.env.DOCKERHUB_USERNAME,
      password: process.env.DOCKERHUB_PASSWORD,
      timeout: 30000
    }
  ]
};
```

Any ideas on this (or am I just missing some obvious setting? 😅)

**Reply by jokay** [reply]
My understanding is, you don't need a runner config.json at all but you can add the Docker Hub credentials by adding the environment variables to your `.gitlab-ci.yml`:

```yml
renovate:
  ...
  variables:
    DOCKER_USERNAME: your_uid
    DOCKER_PASSWORD: your_pwd
  ...
```

**Reply by jokay** [reply]
In fact for your case there may be a typo `DOCKERHUB_USERNAME` vs `DOCKER_USERNAME`.

**Reply by malmor** [reply]
`DOCKERHUB_*` refers to our private/internal registry, while `DOCKER_PUBLIC_REGISTRY_*` refers to hub.docker.com.
Authentication for our private registry works just fine, so the env vars should be present and the host rules should work.

Maybe `matchHost: "https://index.docker.io/"` shouldn't be set? (Leaving it out produces some "HTTP 401 Wrong username/password" errors when querying docker.io, that's why I configured it)

**Reply by rarkins** [reply]
try `matchHost=docker.io` so that it also captures `auth.docker.io`

**Reply by rarkins** [reply]
If you're only using Docker Hub then you could leave `matchHost` out completely and rely on `hostType` alone

**Reply by malmor** [reply]
Hey @rarkins,
thanks for the suggestions.

We are running a private container registry, so I can't leave out the `matchHost`. I simplified the configuration to `matchHost=docker.io`, but we are still facing those "too many request" errors. (We are monitoring our remaining requests for our docker.io user - that limit is definitely not reached)

I took a more detailed look into the debug logs and found this block ("Combined config"):
```
...
{"hostType":"docker","matchHost":"docker.io","username":"secret_docker_io_user","password":"***********","timeout":30000},
{"hostType":"docker","matchHost":"registry.company.local,"username":"secret_private_registry_user","password":"***********","timeout":30000},
{"hostType":"docker","matchHost":"env.ci.runner.short","token":"***********"},
{"hostType":"docker","matchHost":"public.registry","username":"secret_docker_io_user","password":"***********"},
{"hostType":"docker","username":"secret_private_registry_user","password":"***********"}
...
```

Lines 1-2 are as expected, 3 is some GitLab CI stuff (?), no idea where 4 comes from and 5 might result from the fact that we have global GitLab CI variables called `DOCKER_USERNAME` and `DOCKER_PASSWORD` containing out private registry credentials.

I will try saving our docker.io credentials as `DOCKER_*` env vars, maybe this is causing some confusion for docker/renovate.
If you have any other thoughts I will be happy to try them 🚀

**Reply by rarkins** [reply]
Seems like these are probably coming from your env. See https://docs.renovatebot.com/getting-started/private-packages/#hostrules-configuration-using-environment-variables

**Reply by rarkins** [reply]
And yes it could explain why that last rule's username/password is being used instead of your first rule's

**Reply by malmor** [reply]
You were right, those hostRules were generated from some environement variables being passed into the container.
I removed them, which resulted in a "clean" hostRules config:
```
"hostRules": [
    {
        "hostType": "docker",
        "matchHost": "docker.io",
        "username": "secret_docker_io_user",
        "password": "***********",
        "timeout": 30000
    },
    {
        "hostType": "docker",
        "matchHost": "registry.company.local",
        "username": "secret_private_registry_user",
        "password": "***********",
        "timeout": 30000
    },
    ... // some maven and npm host rules
]
```

Sadly it didn't change anything - the "too many requests" errors still occur 😕.
Are there any advanced settings/log levels (currently debug)/... that I can use to troubleshoot this?

I'd love to get this fixed - Renovate is a fantastic tool, but those pull rate limit errors are really annoying 😄

**Reply by viceice** [reply]
Can you provide new debug logs please

**Reply by malmor** [reply]
Hey, sure!

Here are our .gitlab-ci.yml ([gitlab-ci.yml.txt](https://github.com/renovatebot/renovate/files/7067523/gitlab-ci.yml.txt)), our renovatebot configuration ([renovate-runner-config.js.txt](https://github.com/renovatebot/renovate/files/7067519/renovate-runner-config.js.txt)) and the debug output containing the too-many-requests error ([renovate-runner-debug-log.txt](https://github.com/renovatebot/renovate/files/7067520/renovate-runner-debug-log.txt)).

I stripped the debug output down to all general output + one failing repo - I hope that works.
Thanks for taking your time helping me/us 👏 👍 

If there is anything I can do to help please let me know!

**Reply by viceice** [reply]
That's strange you get blocked by 28 requests? I've never seen this before.

```json
[
  "auth.docker.io, 11 requests, 460ms request average, 0ms queue average",
  "gitlab.company.local, 4 requests, 83ms request average, 0ms queue average",
  "index.docker.io, 28 requests, 465ms request average, 0ms queue average"
]
```

How often are you running renovate?

**Reply by malmor** [reply]
Hey @viceice, the attached log just contains one failing repository - there are a few more docker.io requests.
I looked into the requests summaries for all 104 processed repositories. They can be summed up as follows:
```
=> renovate started

auth.docker.io: 36 requests
index.docker.io: 90 requests

=> HTTP Error: rate limit reached ("getManifestResponse")

auth.docker.io: 6 requests
index.docker.io: 14 requests

=> HTTP Error: rate limit reached ("getManifestResponse")

auth.docker.io: 2 requests
index.docker.io: 2 requests

=> HTTP Error: rate limit reached ("getManifestResponse")

auth.docker.io: 1 request
index.docker.io: 4 requests

=> renovate finished
```

Renovate is running once per hour (takes ~25min to complete). But our hub.docker.com user is being used on some other systems aswell (docker login). Looking at those numbers makes me wonder if we might actually be reaching the pull rate limit (which is at 200 requests per 6 hours) and that there is no misconfiguration going on with our Renovate at all.
(But on the other hand we have a monitoring set up for the remaining pull rate limit, based on https://docs.docker.com/docker-hub/download-rate-limit/ and https://about.gitlab.com/blog/2020/11/18/docker-hub-rate-limit-monitoring/, which is showing no problems at all 🤔)

**Reply by viceice** [reply]
I think #11473 will fix your rate limit issue

**Reply by malmor** [reply]
Great idea 🥇 

This would be really useful as `HEAD` requests are not counted for the hub.docker.com rate limit 👍 
https://docs.docker.com/docker-hub/download-rate-limit/#definition-of-limits

**Reply by viceice** [reply]
I know 😉
